New Login Pattern

yih-wang opened the pull request:
Update login pattern with new templates and additional guidelines.

Fix issue #406

yih-wang commented:
Updated the Login Pattern to include new login templates and additional guidelines of error messages. However, the document seems too large for a single pattern in PatternFly. I think it would be better to divide it to separate patterns if possible.

junezhang commented:
@mcarrano @LHinson @catrobson please help to review this PR.

mcarrano commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+3. Social media buttons should line up in one column.
+
+## Error Messages
+User input can be validated on the client and on the server. Thus common error conditions can be divided to two aspects. Jump to [Client-side Errors](#client-side-errors) or [Server-side Errors](#server-side-errors)
```
I'm not convinced that this distinction is important for specifying what and how the messages are returned. It seems like the only difference from a user perspective is whether validation can happen after the field loses focus or must wait for the user to click the submit button. The former would only apply to thinks like formatting errors (i.e. invalid characters that don't require a call to the backend authenticator). Could we consolidate this section a bit?

yih-wang replied:
I added some explanations to clarify the features of client-side & server-side validations, like how and when to display error messages. Please have review.

mcarrano commented on pattern-library/application-framework/login-page/design/overview.md:
```diff
@@ -11,5 +11,22 @@ The login page allows the user to gain access to an application by entering thei
 
-## Example Login Page
-![Image of login page](img/login-01.png)
+Jump to [Basic Login Page](#basic-login-page), [Identity First](#identity-first), [Social Login](#social-login), [SSO Login](#sso-login) or [Other Flows](#other-flows)
```
It would be helpful to see a sentence or two at the start of each of these sections that explains the usage for each of these login patterns (i.e. how used and when used).

yih-wang replied:
Fixed that. Thanks for all the comments!

mcarrano commented on pattern-library/application-framework/login-page/design/overview.md:
```diff
@@ -11,5 +11,22 @@ The login page allows the user to gain access to an application by entering thei
+## SSO Login
+![Image of sso login](img/SSO-1.png)
+
+## Other Flows
```
Aren't these just examples of Two-Factor Authentication? Is that what we should call this?

yih-wang replied:
According to this comment (https://docs.google.com/a/redhat.com/document/d/1CdxlqxskfpjMW_i5AS8KfdcH7Sr2cNutMAThw_Sj_n0/edit?disco=AAAAA9i2STw), one login process could have any number of steps. So I'm not sure if it is proper to call them "Two-Factor".

junezhang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+6. **Help Link**: It is a link for users to recover their password.
+7. **Login Button**: The button name is better to be called “Log In”. According to the Terminology and Wording Style Guide, it is better to use a verb phrase (log in) rather than a noun or adjective phrase (login) in terms of button name.
+8. **Sign up**: The page will jump to the sign up flow after clicking the “Sign up” link.
+9. **Language Selector** (optional): The language selector is located under the login card.
```
@mcarrano @serenamarie125  @catrobson This is how the downstream usage: https://redhat.invisionapp.com/share/VTBZSUVS7#/screens/236959445
How do you think of putting the language selector on the top of the login card.

zhutaoredhat commented:
@mcarrano You said it.  It makes sense to break into some smaller patterns, we had discussed this idea either, but there is one thing not sure whether put them all under the "Application Framework" or we need to create a new tab called "Login Page". Thanks a lot.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+
+### Basic Login Page
+![Image of basic login page](img/login-page-1.png)
+1. **Logo**: The product logo is placed at the top.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+3. **Title**: The title is located at the top of the login card.
+4. **Required Fields**: Username and password are the required fields while additional fields are optional. Some applications may require additional fields such as a specific server or geo-location. When additional fields are required, they can appear underneath the password field. The login card will expand to accommodate the additional fields.
+The additional fields order will be addressed on a case-by-case basis, but they should sit between the “Password field” and the “checkbox”. Try to limit the number of the fields to less than five. Also, you may use progressive disclosure if the field value have dependency on the previous relevant field.
+5.  **Checkbox** (optional): Checkbox is located under the Required fields. For security reasons, we do not recommend allowing users to remember their password. Checkbox labels should be capitalized according to [Terminology and Wording Style Guide](http://www.patternfly.org/styles/terminology-and-wording/).
```
Agree. Modified this rule and clarify the label can be customized.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+4. **Required Fields**: Username and password are the required fields while additional fields are optional. Some applications may require additional fields such as a specific server or geo-location. When additional fields are required, they can appear underneath the password field. The login card will expand to accommodate the additional fields.
+The additional fields order will be addressed on a case-by-case basis, but they should sit between the “Password field” and the “checkbox”. Try to limit the number of the fields to less than five. Also, you may use progressive disclosure if the field value have dependency on the previous relevant field.
+5.  **Checkbox** (optional): Checkbox is located under the Required fields. For security reasons, we do not recommend allowing users to remember their password. Checkbox labels should be capitalized according to [Terminology and Wording Style Guide](http://www.patternfly.org/styles/terminology-and-wording/).
+6. **Help Link**: It is a link for users to recover their password.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+The additional fields order will be addressed on a case-by-case basis, but they should sit between the “Password field” and the “checkbox”. Try to limit the number of the fields to less than five. Also, you may use progressive disclosure if the field value have dependency on the previous relevant field.
+5.  **Checkbox** (optional): Checkbox is located under the Required fields. For security reasons, we do not recommend allowing users to remember their password. Checkbox labels should be capitalized according to [Terminology and Wording Style Guide](http://www.patternfly.org/styles/terminology-and-wording/).
+6. **Help Link**: It is a link for users to recover their password.
+7. **Login Button**: The button name is better to be called “Log In”. According to the Terminology and Wording Style Guide, it is better to use a verb phrase (log in) rather than a noun or adjective phrase (login) in terms of button name.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+5.  **Checkbox** (optional): Checkbox is located under the Required fields. For security reasons, we do not recommend allowing users to remember their password. Checkbox labels should be capitalized according to [Terminology and Wording Style Guide](http://www.patternfly.org/styles/terminology-and-wording/).
+6. **Help Link**: It is a link for users to recover their password.
+7. **Login Button**: The button name is better to be called “Log In”. According to the Terminology and Wording Style Guide, it is better to use a verb phrase (log in) rather than a noun or adjective phrase (login) in terms of button name.
+8. **Sign up**: The page will jump to the sign up flow after clicking the “Sign up” link.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+
+### Identity First
+![Image of identity first](img/login-page-2.png)
+1. **Username Field**: Username should be the first input and independently validated by server. Some username may not need password. When users click the “Next” Button, the page will  jump to the relevant application directly. In the situation that usernames do not exist, an error message will pop up after user clicking the “Next Button”.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+### Identity First
+![Image of identity first](img/login-page-2.png)
+1. **Username Field**: Username should be the first input and independently validated by server. Some username may not need password. When users click the “Next” Button, the page will  jump to the relevant application directly. In the situation that usernames do not exist, an error message will pop up after user clicking the “Next Button”.
+2. **Next Button**:The button should be called “Next” instead of “Log In”. Because most users still need to input password as a second step.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+
+### Social Login
+![Image of social login](img/login-page-4.png)
+1. **Social Media Buttons**: When button number less than four, all buttons will be displayed in one column. On the contrary, buttons will always be displayed in two columns.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+
+#### Field(s) is empty
+![Image of empty fields](img/Message-1.png)
+1. **When to Show**: When user clicks the Action button (Log in/Next), the login form won’t be submitted and error message(s) will show below the empty field(s) if fields are still empty, .
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+#### Field(s) is empty
+![Image of empty fields](img/Message-1.png)
+1. **When to Show**: When user clicks the Action button (Log in/Next), the login form won’t be submitted and error message(s) will show below the empty field(s) if fields are still empty, .
+2. **When to Hide**: If user fill the empty field, the error message will disappear when text field loses focus.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,114 @@
+
+![Image of identity first](img/login-page-3.png)
+1. **Validated Username**: After successful verification, the username should change to  read-only mode in this page.
+2. **Edit**: Users can change username by clicking the edit icon on the far right. After clicking the icon, the card will return to the text field mode as the previous page.
```
In fact, clicking the edit icon will also lead user back to step one. The only difference (between the “returned step one" and the "original step one") is the username string is in the username field.

ssilvert commented:
One thing I noticed on several of these screens.  They assume that you would point out if a user name is not found.  While I love the design showing the error message right below the field that is in error, the message itself is considered bad practice.  

You don't want an attacker to be able to find out which user names are valid.  So the error message in this case should always pertain to both, saying something like, "User name and/or password are incorrect".

If the user leaves the user name blank, then it's fine to display the error on the user name field.  So maybe you could keep the design you have but just change the example error message to something like, "User name is required."

junezhang commented:
@ssilvert Thanks for your feedback from the developer perspective.
In this PR, @yih-wang Haley provided two solutions for that the account or password is wrong.
One is directly to tell user that account couldn't be found  or password is wrong. the benefit of that is easier for user to understand which is wrong.
The other is based on the security reason, the error message pertain to both, like you suggested, "Your username or password is incorrect."
SO the application can based on the security level to choose to use whichever is the best fit.

yih-wang commented:
@ssilvert Agree with @junezhang . And you will find some applications like Facebook and Quora choose to point out whether the username or password is wrong. As PatternFly provides public guidelines, I think we should cover both methods to meet different needs. What's your thoughts?

mcarrano commented:
@yih-wang and @junezhang , today @LHinson and I discussed here's how we propose to organize this content:

Login Page (this updates the current login pattern)
 - Standard Login
 - Social login variation
 - Errors and validation

Single Sign-On (new pattern)
 - SSO login

Identity First Login (new pattern)
 - Combine identity first and other login methods into a single pattern since these are very similar (i.e. they all require a multistep login process). Maybe call this "multi-step login?"

This approach will keep the information relevant to most of our products in the current/standard login page and pull out less common patterns into separate documents. All of these should live within the Application Framework section on the site. These patterns should be cross linked as applicable to make sure sures coming to the standard login page can get to other related patterns.

Because of the scope, I'm thinking that it would make the most sense to split this into multiple PRs, one for each pattern. That would mean pulling Identity First and SSO out of the current login page and adding them later via separate PRs. Do you agree? I'm flexible on this point.

Let me know if this overall proposal makes sense and we can move forward. If you have questions, will be glad to schedule some time to discuss.

catrobson commented:
@mcarrano that breakdown makes sense to me

junezhang commented:
@mcarrano thanks for your updating. we all think the breakdown makes a lot more sense.  Currently we still have a round review of the wireframe, when finished the wireframe updating, we will update the PR as you suggested.

yih-wang commented:
@mcarrano @junezhang I have separated the document into three PRs. They still need updates based on new comments before the final review.

yih-wang commented:
@mcarrano I have updated all the three PRs. The other two are #503 and #504 . 

I changed the name of multiple steps login to "Multi-Factor Login". And for the 2nd step of Identity First, I kept both the language selector and sign up link in the screen to keep the layout consistent with the 1st step, although user seldom use these two functions in this scenario. 

Please have a review. Thanks!

mcarrano commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+3. **Recommended Message**: "Caps lock is currently on. This may cause you to type an incorrect password."
+
+### Server-side Errors
+When users submit the form, the entire page is reloaded with indications of validation errors. At the same time, the password field should be cleared due to the security reason. Error messages can perform as inline validations for each field or inline notifications as global messages. Inline Notifications are displayed below the application name and above the input fields. If there are multiple m ...
```
Change to... "When users submit the form, the entire page is reloaded if there are validation errors. At the same time, the password field should be cleared due to the security reason. Error messages can appear as inline errors for each field or as inline notifications. Inline Notifications are displayed below the application name and above the input fields. If there are multiple messages, each message appears on its own line, and the area adjusts as needed to fit these messages."

yih-wang replied:
Fixed.

mcarrano commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+#### Account doesn’t exist/password is wrong
+If the user input a wrong account name which do not exist or if the user input a wrong password, we provide two solutions to show error messages:
+
+1. Directly tell the user that the account doesn’t exist or the password is wrong. Although it is useful and clear for user to easily understand the problem, it’s not safe.
```
Change to... "Directly tell the user that the account doesn’t exist or the password is wrong. Although this approach is useful and clear for user, it's not recommended if protecting identity of valid users is important."

yih-wang replied:
Fixed.

mcarrano commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+  - "The account doesn’t exist. Try again."
+  - “The account doesn’t exist. Try again or Sign up for PatternFly.” for the products supporting signing up. The “Sign up for PatternFly” should be a link to help users creating a new account more easier. “PatternFly” can be replaced with other product names.
+ - When password is wrong, we recommend to show “Incorrect password, please try again.”
+If the user input an old password (they may have changed the password recently), we can show an error message alerting them that the password has changed instead of stating “the password is wrong” only.
```
"inputs" not "input"

yih-wang replied:
Fixed.

mcarrano commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+#### Other Examples of Error Messages
+![Image of other error messages](img/Server-4.png)
 
 ## Notifications
```
I was unclear why these informational messages are shown off to the side rather than inline like other error messages. As you can see from http://www.patternfly.org/pattern-library/communication/inline-notifications/ you can put a close button on the informational alerts if you want to allow users to close them. I think that keeping one alert/notification style will simplify this.

yih-wang replied:
I used to put it in the middle under the logotype. But I saw it is shown off to the side as a toast notification in the implemented mockup (https://rawgit.com/matthewcarleton/patternfly/login-responsive-dist/dist/tests/login-basic.html). Of course it's better to use an inline notification to keep consistency. The only concern is it will take up another row as our vertical space is precious. Is it possible for a notification floating on top of the logo (centered in screen)? I'm not sure which pattern it could inherit..

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+7. **Help Link** (optional): It is a link for users to recover their password.
+8. **Login Button**: The button should be labeled "Log In". See the [Terminology and Wording Style Guide](http://www.patternfly.org/styles/terminology-and-wording/) for more information.
+9. **Sign up** (optional): The page will jump to the sign up flow after clicking the “Sign up” link.
+10. **Links**: The links are entrance for users to get information and helps.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+### Social Login
+![Image of social login](img/login-page-4.png)
+1. **Social Media Buttons**: When the number of buttons is four or less, display all buttons in one column. Otherwise, use a two-column layout.
+2. **More Button**: When there are more than 8 buttons, just show first 7 of them. User can click the More Button to expand the list and view all the buttons.
```
It will directly show the buttons. I added a screen here to clarify the behavior.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+This section is talking about how the login pages look on small screens.
+
+![Image of responsive state](img/responsive.png)
+1. When the screen getting smaller, the help link will automatically moved to the next line if there isn’t enough space for checkbox and help link placed in the same line.
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+2. Social media buttons follow the same rule as in desktop size.
+
+## Error Messages
+User input can be validated on the client and on the server. With client-side validation, error message will be presented before form submitted. It could happen when the text field gains focus, loses focus, or when user clicks the action button (without calling backend authenticator). With server-side validation, error message will only be presented when the page is reloaded with the data user su ...
```
Fixed.

yih-wang commented on pattern-library/application-framework/login-page/design/design.md:
```diff
@@ -1,35 +1,89 @@
+Jump to [Client-side Errors](#client-side-errors) or [Server-side Errors](#server-side-errors)
+
+### Client-side Errors
+In most cases, client-side validations perform as inline validations for each field. Some common conditions are as follows.
```
Fixed.

yih-wang commented:
@mcarrano I have updated the PRs as we discussed. Please have a review when you are available. Great thanks!

mcarrano commented:
@yih-wang I have reviewed your latest changes and everything looks good. Thanks for all your iterations on these! I am going to merge this PR so we get these changes into our design library. If anyone has further issues, we can address via a follow-up PR. We also should update all mockups with hi-res versions.
